File: src/main/java/alfheim/common/core/command/CommandAlfheim.java
package alfheim.common.core.command;

import alexsocol.asjlib.ASJUtilities;
import alfheim.AlfheimCore;
import alfheim.common.core.handler.*;
import alfheim.common.core.registry.AlfheimRecipes;
import alfheim.common.core.util.AlfheimConfig;
import alfheim.common.integration.thaumcraft.ThaumcraftAlfheimModule;
import alfheim.common.network.Message3d;
import alfheim.common.network.Message3d.m3d;
import net.minecraft.command.*;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.server.MinecraftServer;
import net.minecraft.util.*;
import vazkii.botania.common.Botania;

import java.util.*;

public class CommandAlfheim extends CommandBase {
	
	@Override
	public int getRequiredPermissionLevel() {
		return 2;
	}
	
	@Override
	public List getCommandAliases() {
		return Arrays.asList("alf");
	}
	
	@Override
	public String getCommandName() {
		return "alfheim";
	}
	
	@Override
	public String getCommandUsage(ICommandSender sender) {
		return "alfheim.commands.alfheim.usage";
	}
	
	@Override
	public void processCommand(ICommandSender sender, String[] args) {
		if(args.length == 1) {
			boolean esmOld = AlfheimCore.enableElvenStory;
			boolean mmoOld = AlfheimCore.enableMMO;
			if (args[0].equalsIgnoreCase("ESM")) toggleESM(AlfheimCore.enableElvenStory = !AlfheimCore.enableElvenStory); else
			if (args[0].equalsIgnoreCase("MMO")) toggleMMO(AlfheimCore.enableMMO = !AlfheimCore.enableMMO); else
			throw new WrongUsageException("alfheim.commands.alfheim.wrong");
			AlfheimConfig.writeModes();
		
			ASJUtilities.sayToAllOnline(String.format(StatCollector.translateToLocal("alfheim.commands.alfheim.done"),
										sender.getCommandSenderName	(),
										AlfheimCore.enableElvenStory? EnumChatFormatting.GREEN : EnumChatFormatting.DARK_RED,
										EnumChatFormatting.RESET	,
										AlfheimCore.enableMMO		? EnumChatFormatting.GREEN : EnumChatFormatting.DARK_RED,
										EnumChatFormatting.RESET	));
			
			AlfheimCore.network.sendToAll(new Message3d(m3d.TOGGLER, args[0].equalsIgnoreCase("ESM") ? 1 : 0, ((esmOld ? 1 : 0) << 1) | (AlfheimCore.enableElvenStory ? 1 : 0), ((mmoOld ? 1 : 0) << 1) | (AlfheimCore.enableMMO ? 1 : 0)));
		} else throw new WrongUsageException("alfheim.commands.alfheim.wrong");
	}
	
	@Override
	public List addTabCompletionOptions(ICommandSender sender, String[] args) {
		return getListOfStringsMatchingLastWord(args, "ESM", "MMO");
	}
	
	public static void toggleESM(boolean on) {
		if (on) {
			AlfheimConfig.initWorldCoordsForElvenStory(AlfheimCore.save);
			EventHandler.checkAddAttrs();
			if (Botania.thaumcraftLoaded) ThaumcraftAlfheimModule.addESMRecipes();
		} else {
			if (Botania.thaumcraftLoaded) ThaumcraftAlfheimModule.removeESMRecipes();
			toggleMMO(AlfheimCore.enableMMO = false);
		}
	}
	
	public static void toggleMMO(boolean on) {
		if (on) {
			CardinalSystem.load(AlfheimCore.save);
			AlfheimRecipes.addMMORecipes();
			toggleESM(AlfheimCore.enableElvenStory = true);
			for (Object o : MinecraftServer.getServer().getConfigurationManager().playerEntityList) CardinalSystem.transfer((EntityPlayerMP) o);
		} else {
			CardinalSystem.save(AlfheimCore.save);
			AlfheimRecipes.removeMMORecipes();
		}
	}
}

File: src/main/java/alfheim/common/core/command/CommandAlfheim.kt
package alfheim.common.core.command

import alexsocol.asjlib.*
import alfheim.AlfheimCore
import alfheim.api.ModInfo
import alfheim.common.achievement.AlfheimAchievements
import alfheim.common.core.handler.*
import alfheim.common.crafting.recipe.AlfheimRecipes
import alfheim.common.integration.thaumcraft.ThaumcraftAlfheimModule
import alfheim.common.network.Message3d
import alfheim.common.network.Message3d.m3d
import net.minecraft.command.*
import net.minecraft.entity.player.EntityPlayerMP
import net.minecraft.server.MinecraftServer
import net.minecraft.util.*
import net.minecraftforge.common.AchievementPage
import vazkii.botania.common.Botania

class CommandAlfheim: CommandBase() {
	
	override fun getRequiredPermissionLevel() = 2
	
	override fun getCommandAliases() = listOf("alf")
	
	override fun getCommandName() = "alfheim"
	
	override fun getCommandUsage(sender: ICommandSender) = "alfheim.commands.alfheim.usage"
	
	override fun processCommand(sender: ICommandSender, args: Array<String>) {
		if (args.size == 1) {
			val esmOld = AlfheimConfigHandler.enableElvenStory
			val mmoOld = AlfheimConfigHandler.enableMMO
			when {
				args[0].equals("ESM", ignoreCase = true) -> {
					AlfheimConfigHandler.enableElvenStory = !AlfheimConfigHandler.enableElvenStory
					toggleESM(AlfheimConfigHandler.enableElvenStory)
				}
				
				args[0].equals("MMO", ignoreCase = true) -> {
					AlfheimConfigHandler.enableMMO = !AlfheimConfigHandler.enableMMO
					toggleMMO(AlfheimConfigHandler.enableMMO)
				}
				
				else                                     -> throw WrongUsageException("alfheim.commands.alfheim.wrong")
			}
			
			ASJUtilities.sayToAllOnline(String.format(StatCollector.translateToLocal("alfheim.commands.alfheim.done"),
													  sender.commandSenderName,
													  if (AlfheimConfigHandler.enableElvenStory) EnumChatFormatting.GREEN else EnumChatFormatting.DARK_RED,
													  EnumChatFormatting.RESET,
													  if (AlfheimConfigHandler.enableMMO) EnumChatFormatting.GREEN else EnumChatFormatting.DARK_RED,
													  EnumChatFormatting.RESET))
			
			AlfheimCore.network.sendToAll(Message3d(m3d.TOGGLER, (if (args[0].equals("ESM", ignoreCase = true)) 1 else 0).D, ((if (esmOld) 1 else 0) shl 1 or if (AlfheimConfigHandler.enableElvenStory) 1 else 0).D, ((if (mmoOld) 1 else 0) shl 1 or if (AlfheimConfigHandler.enableMMO) 1 else 0).D))
		} else
			throw WrongUsageException("alfheim.commands.alfheim.wrong")
	}
	
	override fun addTabCompletionOptions(sender: ICommandSender?, args: Array<String>?) = getListOfStringsMatchingLastWord(args, "ESM", "MMO")
	
	companion object {
		
		fun toggleESM(on: Boolean) {
			if (on) {
				AlfheimConfigHandler.initWorldCoordsForElvenStory(AlfheimCore.save)
				ESMHandler.checkAddAttrs()
				AchievementPage.getAchievementPage(ModInfo.MODID.capitalize()).achievements.add(AlfheimAchievements.newChance)
				if (Botania.thaumcraftLoaded) ThaumcraftAlfheimModule.addESMRecipes()
			} else {
				if (Botania.thaumcraftLoaded) ThaumcraftAlfheimModule.removeESMRecipes()
				AlfheimConfigHandler.enableMMO = false
				toggleMMO(AlfheimConfigHandler.enableMMO)
				AchievementPage.getAchievementPage(ModInfo.MODID.capitalize()).achievements.remove(AlfheimAchievements.newChance)
			}
		}
		
		fun toggleMMO(on: Boolean) {
			if (on) {
				CardinalSystem.load(AlfheimCore.save)
				AlfheimRecipes.addMMORecipes()
				AlfheimConfigHandler.enableElvenStory = true
				toggleESM(AlfheimConfigHandler.enableElvenStory)
				for (o in MinecraftServer.getServer().configurationManager.playerEntityList) CardinalSystem.transfer(o as EntityPlayerMP)
			} else {
				CardinalSystem.save(AlfheimCore.save)
				AlfheimRecipes.removeMMORecipes()
			}
		}
	}
}